File: C/cplus/algorithm/graph.cc
#include <iostream>

using namespace std;

class Line
{
public:
	Line(int e,int w):end(e),weight(w),next(NULL){};
	~Line(){};
	int end;
	int weight;
	Line *next;
};

class vertex
{
public:
	vertex(int s):start(s),number(0),neighbor(NULL),next(NULL){};
	~vertex(){};
	int start;
	int number;
	Line *neighbor;
	vertex *next;
};

class Graph
{
public:
	Graph(int s, int e, int w){
		count = 1;
		head = create_new_vertex_for_graph(s,e,w);
	};

	vertex* create_new_vertex_for_graph(int s, int e, int w){
		vertex *v = new vertex(s);
		v->neighbor = new Line(e,w);
		return v;
	};
	vertex* find_vertex_in_graph(vertex* v, int s){
		if(NULL==v)
			return NULL;

		while(v){
			if(v->start == s)
				return v;
			v = v->next;
		}
		return NULL;
	};
	Line* find_line_in_graph(Line* l, int e){
		if(NULL==l)
			return NULL;
		while(l){
			if(l->end == e)
				return l;
			l = l->next;
		}
		return NULL;
	};
	bool insert_vertex_into_graph(int start, int end, int weight){
		Vertex* v=NULL;
		Line* l=NULL;
		if(NULL==head){
			head = create_new_vertex_for_graph(start, end, weight);
			head->number++;
			count++;
			return TRUE;
		};

		v = find_vertex_in_graph(head, s);
		if(v==NULL){
			v = create_new_vertex_for_graph(start, end, weight);
			v->next = head;
			head->next = v;
			head->number++;
			count++;
			return TRUE;
		};

		l = find_line_in_graph(v->neighbor, end);
		if(l) return FALSE;
		l = new Line(end, weight);
		l->next = v->neighbor;
		v->neighbor = l;
		v->number++;
		return TRUE;
	};

	bool delete_old_vertex(Vertex* v, int s){

	};

	bool delete_old_line(Line* l, int e){
	};

	bool delete_vertex_from_graph(int start, int end, int weight){

	};
	~Graph(){};
	int count;
	vertex *head;
};

int main(int argc, char const *argv[])
{
	
	return 0;
}

File: Go/graph.go
package main

import (
    "fmt"
)

// 1. Matrix
var graph_matrix [5][5]int

// 2. Link
type Line struct {
    end        int
    weight     int
    next       *Line
}

type Vectex struct {
    start       int
    number      int
    neighbor    *Line
    next        *Vectex
}

type Graph struct {
    count       int
    head        *Vectex
}

func create_new_line(end int, weight int) *Line {
    return &Line {
        end: end,
        weight: weight,
        next: nil,
    }
}

func create_new_vectex(start int) *Vectex {
    return &Vectex {
        start: start,
        number: 0,
        neighbor: nil,
        next: nil,
    }
}

func create_new_vectex_for_graph(start int, end int, weight int) *Vectex {
    v := create_new_vectex(start)

    v.neighbor := create_new_line(end, weight)
}

func create_new_graph(start int, end int, weight int) *Graph {
    graph := new(Graph)

    graph.count = 1
    graph.head = create_new_vectex_for_graph(start, end, weight)
}

func find_vectex_in_graph(v *Vectex, start int) *Vectex {
    if v == nil {
        return nil
    }

    for v!=nil {
        if start == v.start {
            return v
        }

        v = v.next
    }

    return nil
}

func find_line_in_graph(l *Line, end int) *Line {
    if l == nil {
        return nil
    }

    for l!=nil {
        if end == l.end {
            return l
        }

        l = l.next
    }

    return nil
}

func insert_vectex_into_graph(graph *Graph, start int, end int, weight int) bool {
    if graph == nil {
        return false
    }

    if graph.head == nil {
        graph.head = create_new_vectex_for_graph(start, end, weight)
        graph.head.number++
        graph.count++
        return true
    }

    // Try to find out if start already exists
    vec := find_vectex_in_graph(graph.head, start)
    if  vec==nil {
        vec = create_new_vectex_for_graph(start, end, weight)
        vec.next = graph.head
        graph.head = vec
        graph.head.number++
        graph.count++
        return true
    }

    line := find_line_in_graph(vec.neighbor, end, weight)
    if line!=nil {
        return false
    }

    line = create_new_line(end, weight)
    line.next = vec.neighbor
    vec.neighbor = line

    return true
}

func delete_old_vectex(start int, pv **Vectex) bool {
    if pv == nil || *pv == nil {
        return false
    }

    v := find_vectex_in_graph(*pv, start)
    if v == nil {
        return false
    }

    if v == *pv {
        *pv = v.next
        return true
    }

    pre := *pv
    for pre.next!=v {
        pre = pre.next
    }

    pre.next = v.next
    return true
}

func delete_old_line(end int, pline **Line) bool {
    if pline == nil || *pline == nil {
        return false
    }

    l := find_line_in_graph(*pline, end)
    if l == nil {
        return false
    }

    if l == *pline {
        *pline = l.next
        return true
    }

    pre := *pline
    for pre.next!=l {
        pre = pre.next
    }

    pre.next = l.next
    return true
}

func delete_vectex_from_graph(graph *Graph, start int, end int, weight int) bool {
    if graph==nil || graph.head==nil {
        return false
    }

    v := find_vectex_in_graph(graph.head, start)
    if v==nil {
        return false
    }

    line := find_line_in_graph(v.neighbor, end)
    if line==nil {
        return false
    }

    res := delete_old_line(end, &v.neighbor)
    v.number--
    if len(v.neighbor) == 0 {
        res = delete_old_vectex(start, &graph.head)
    }

    graph.count--
    return res
}

func main() {
    fmt.Println(graph_matrix)
}